Check for and install missing core dependencies

settermjd opened the pull request:
## Description

Checks for and installs any missing [core dependency](https://doc.owncloud.org/server/latest/admin_manual/installation/source_installation.html#install-the-required-packages).

**This is currently a WIP!**

## Motivation and Context

After a series of times where I've been frustrated that packages were missing and then had to consult the documentation to find what they were, and for a simple script to install them, I wrote this small script to do it all programmatically. Currently, it only supports Ubuntu 17.10, but, I plan to improve upon it for future releases.

## How Has This Been Tested?

I tested the script on a local installation.

## Screenshots (if appropriate):

n/a

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

settermjd committed:
Check for missing core dependencies and install those that are missing

After a series of times where I've been frustrated that packages were
missing and then had to consult the documentation to find what they
were, and for a simple script to install them, I wrote this small script
to do it all programmatically. Currently, it only supports Ubuntu 17.10,
but, I plan to improve upon it for future releases.

mmattel commented:
Have the webserver and the database as choice (apache, nginx, mysql, maria-db)

patrickjahns commented on bin/check-dependencies.sh:
```diff
@@ -0,0 +1,101 @@
+function check_distribution()
+{
+  distribution=$( grep -i "version_id" /etc/os-release | awk -F'=' '{print $2}' | tr -d '"')
+  if [ $distribution != "17.10" ]
```
Please propose a LTS version - that is what we officially support

settermjd replied:
Sounds fair. Thanks for the feedback.

settermjd replied:
@patrickjahns, the script now supports Ubuntu 17.04, 16.04, and 16.04.4.

settermjd replied:
https://github.com/owncloud/core/pull/31078/commits/cde28ba481f4e6e7668216bf2e04e79974ed75a7 added support for the previous LTS release, 14.04.

settermjd committed:
Improved the dependency check and install script

This commit allows for other distributions (and versions) to be
supported, requiring only that they be added to the supported
distributions list and that a dependencies config file be created in the
dependencies directory.

Other than that:

- the script's output is more meaningful and easier to read
- apt output is stored in a log file
- the script is now easier to maintain

settermjd committed:
Added Ubuntu 16.04 and 16.04.4 as supported distributions

settermjd committed:
Fixed bug in checking if a distro is supported or not

Previously, false positives were returned when checking if a
distribution was supported or not. This has been fixed by reverting to a
simple array loop. Sure, there are "fancier" ways, I'm sure, but
sometimes it's the simple things in life that are often the best. :-)

settermjd committed:
Minor code reorganisation

settermjd committed:
Fixed rendering of Unicode chars in BSD shell

settermjd committed:
Add support for Ubuntu 14.04

settermjd committed:
Add a note about watching/debugging the package installation process

As the APT output is sent to the log file, then there will be nothing to
see, unless that file's tailed. So, it won't be obvious what's gone
wrong, during this aspect of the script, if something has. So this
note's added, just as a simple way of helping the user out.

settermjd committed:
Removing redirection of stdout and stderr to log file

During testing on Ubuntu 14.04, the installation never completed, as
installing MariaDB asked for a password. Given that, the redirection is
being removed.

heiglandreas commented on bin/check-dependencies.sh:
```diff
@@ -0,0 +1,167 @@
+##
+function get_distribution()
+{
+  version=$( grep -E "^VERSION_ID" /etc/os-release | awk -F'=' '{print $2}' | tr -d '"' )
```
Will `/etc/os-release` be available on every target-system?

heiglandreas replied:
I'd rather see whether I can get that information from `/proc` or similar…

settermjd replied:
I'm doing a bit more digging, but I believe so.

jnweiger replied:
to avoid awk ugliness, you can use 
```version=$(. /etc/os-release; echo $VERSION_ID)```

settermjd replied:
That's pretty wicked @jnweiger. I'd never thought of doing that, _until now_. Thanks!!

heiglandreas commented on bin/check-dependencies.sh:
```diff
@@ -0,0 +1,167 @@
+done
+shift $((OPTIND - 1))
+
+if [ $( whoami ) != "root" ]
```
I'd set that right at the beginning of the script. What sense does it make to call `script -h` only to then see that running `script do-something-sensible` doesn't work because it needs to be run as root. But that's just my personal view ;-)

settermjd replied:
Good point. I'd not thought of that. Thanks for the feedback.

heiglandreas reviewed (commented):
I assume that the syntax is working and as you requested it to be run in bash that should be no problem ;)

jnweiger commented on bin/check-dependencies.sh:
```diff
@@ -0,0 +1,167 @@
+
+  # Determine the correct dependencies configuration file to read in
+  distro_lower=$( echo $(get_distribution) | awk '{ gsub(" ", "-"); print tolower($0) }' )
+  source "bin/dependencies/${distro_lower}.cfg"
```
avoiding another awk:
`distro_lower=$( echo $(get_distribution) | tr 'A-Z ' 'a-z-' )`

jnweiger commented on bin/check-dependencies.sh:
```diff
@@ -0,0 +1,167 @@
+    if dpkg -l ${pkg} 2>/dev/null | grep -E -q "^ii"  
+    then
+      echo -e " \xE2\x9c\x94 ${pkg} is installed"
+    else
```
Please add a comment, what \xE2\x9c\x94  means. A unicode glyph on some terminals?

jnweiger commented on bin/check-dependencies.sh:
```diff
@@ -0,0 +1,167 @@
+## 
+## This script checks if the required dependency packages have been installed
+## If one or more have not been installed, then it installs them.
+## It currently only works on Ubuntu 17.10 and has to be run as root.
```
if you write all the `apt` calls as `sudo apt` then it will prompt for the root password if needed (and only once).
The whoami check below should just print a friendly message then, saying that we will prompt for the root password to install stuff....

settermjd replied:
I decided to just remove the `whoami` check altogether.

jnweiger commented on bin/check-dependencies.sh:
```diff
@@ -0,0 +1,167 @@
+## Author: Matthew Setter <matthew@matthewsetter.com>
+##
+
+SUPPORTED_DISTRIBUTIONS=( 
```
comment to add here:
`# CAUTION: keep in sync with the available bin/dependencies/*.cfg files`

settermjd replied:
I'm going to remove this and just check the available config files. It makes no sense to me to check twice when once will suffice.

jnweiger commented on bin/dependencies/ubuntu-14.04.cfg:
```diff
@@ -0,0 +1,20 @@
+core_dependencies=(
+  apache2 mariadb-server 
```
Please update documentation at
https://doc.owncloud.com/server/10.0/admin_manual/installation/system_requirements.html#officially-recommended-supported-options and
https://doc.owncloud.com/server/10.0/admin_manual/installation/source_installation.html
 * There are many more platforms listed there, than we have here. (Tumbleweed should not be a supported platform, imho)
 * The package lists in source_installation.html  differ from the package lists here.
 * The existance of this tool should be mentioned in the docs.
 * This tool should print out the url to the online-docs when started.

jnweiger replied:
@crrodriguez please check the package dependencies against these. Maybe we can interpolate these bin/dependencies/$PLATFORM.cfg lists into spec files and control files?

settermjd replied:
I removed any references to mariadb. Your suggestion makes perfect sense. Thanks. @mmattel, I still need to figure out an NGINX option. Working on that.

settermjd commented:
Need to add the option to install install dependencies, such as wget, make, and git.

settermjd committed:
Add support for Ubuntu 18.04

As noted in the script's comments, it needs to be further refactored so
that it's not looking in two places for whether it supports a given
distro or not. One place should be sufficient.

settermjd committed:
Simplified checks for supported distributions

Previously, the list of distributions was checked in two places, one in
an array in the script, the other in a set of configuration files. This
made no sense, as if the configuration file's there, then it's
(technically) supported. So the check (and need to maintain) in the
array was removed, in favour of the config files.

settermjd committed:
Removed unnecessary AWK calls

While there's nothing wrong with it, it adds unnecessary overhead, when
bash can provide this functionality natively.

settermjd committed:
Removed magic variables for tick and cross in console output

settermjd committed:
Removed unnecessary whoami check

This wasn't needed as the applicable commands were invoked with sudo.
Thanks Juergen.

settermjd committed:
Add further function documentation

mmattel commented:
We have just defined in the documentation for 10.0.8 that the recommendation for php is 7.1
For Ubuntu 14.04 and 16.04 you need to enable a ppa to do so.
```
sudo add-apt-repository -y ppa:ondrej/php
sudo apt-get update -y
```
Then you can install 7.1
I would not recommend to do something different than we do in the documentation :see_no_evil:

settermjd committed:
Updated ubuntu distro dependencies

PVince81 commented:
who's the target audience for this script ? is it for developers ?

what about people who deploy on servers, could they also benefit from this ?

in any case it feels like such script would belong in some part of the documentation (attachment?) or in the [administration repo](https://github.com/owncloud/administration) which already provide various tools for setting up environments

@settermjd

PVince81 commented:
the docs could then be adjusted to point at a subfolder on the administration repo if needed

settermjd commented:
@PVince81, my original intent was to make it simple for anyone to use, not just admins. So, perhaps you're right, that it should go in the docs, and not here in core.

DeepDiver1975 commented:
As discussed in https://github.com/owncloud/core/issues/35777 the master branch will from now on hold the ownCloud 10 codebase. 

This PR targetted ownCloud 11 which is postponed to a far distant future.

Because of that I'm closing this PR and kindly ask you to re-submit this PR in a few days.

Thanks a lot for your patience
